src/data_structures/arguments.rs

```diff
@@ -336,13 +329,17 @@ impl<'a> ArgumentCache<'a> {
     pub fn get_or_compute_coset_evals(
         &mut self,
         coset_idx: usize,
-    ) -> CudaResult<&GenericArgumentStorage<CosetEvaluations>> {
+    ) -> CudaResult<Rc<GenericArgumentStorage<CosetEvaluations>>> {
         assert!(coset_idx < self.used_lde_degree);
 
         if REMEMBER_COSETS == false || coset_idx >= self.fri_lde_degree {
+            let mut tmp_coset = GenericArgumentStorage::allocate(
+                self.monomials.layout.clone(),
+                self.monomials.domain_size(),
+            )?;
             self.monomials
-                .into_coset_eval(coset_idx, self.used_lde_degree, &mut self.tmp_coset)?;
-            return Ok(&self.tmp_coset);
```

Comment:
Hmm, if i'm reading correctly, returning `Ok(&self.tmp_coset)` means if the user calls `get_or_compute_coset_evals` twice to get (references to) two cosets without using them yet, like
```rust
let r0 = cache.get_or_compute_coset_evals(0)?;
let r1 = cache.get_or_compute_coset_evals(1)?;
```
then `r0` and `r1` both reference `tmp_coset`'s memory. The user probably expects r0 to give the first coset's values and r1 to give the second coset's values, but they both give the last-written second coset's values, a nasty silent surprise.

Your approach gives a safer API in addition to avoiding persistent storage for `tmp_coset`.

Reply:
Temporary allocation is utilized only when certain conditions are met, such as when `REMEMBER_COSETS=false` (recompute feature) or when there's a higher degree in the quotient computation.

Reply:
I understand, but I still believe the gotcha can happen under those circumstances.
Anyway, this PR should fix it.

Reply:
> Hmm, if i'm reading correctly, returning `Ok(&self.tmp_coset)` means if the user calls `get_or_compute_coset_evals` twice to get (references to) two cosets without using them yet, like
> 
> ```rust
> let r0 = cache.get_or_compute_coset_evals(0)?;
> let r1 = cache.get_or_compute_coset_evals(1)?;
> ```
> 
> then `r0` and `r1` both reference `tmp_coset`'s memory. The user probably expects `r0` to give the first coset's values and `r1` to give the second coset's values, but they both give the last-written second coset's values, a nasty silent surprise.
> 
> Robert's new approach fixes this API gotcha, in addition to avoiding persistent storage for `tmp_coset`.

that is correct